File: backend/app/schemas/operational_resolution.py
"""
Phase 3A — Operational Resolution Pipeline schemas.

Tightens the operational vocabulary the future production pipeline
will speak:

    Document / batch
            ↓
    Selected or matched invoice pattern
            ↓
    Extracted facts / catalog hints / document metadata
            ↓
    ResolverInput (Phase 2A bridge or direct)
            ↓
    Import Template Resolver (existing dry_run service)
            ↓
    OperationalResolutionResult  ← THIS MODULE

This is intentionally distinct from Phase 2B's "test runner" surface:

  * Phase 2B = test lab. Diagnostic, scenario-driven. Operator UX.
  * Phase 3A = production-shaped pipeline. Same resolver underneath
    but with operational naming, document/batch context, and a
    ``diagnostic_only`` hard-True flag so today's behaviour is
    explicit.

What Phase 3A is NOT:

  * NOT the production export engine.
  * NOT a Review Queue record creator.
  * NOT a document/batch mutator.
  * NOT an OCR / AI extraction surface.

The shape stays close to Phase 2B's ``TemplatePatternTestResult`` so
later phases can promote operational results into Review Queue items
or export-ready rows without re-shaping the contract.
"""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.import_resolver import ExtractedFact, ResolverInput, ResolverResult


# ---------------------------------------------------------------------------
# Request body
# ---------------------------------------------------------------------------


PatternSelectionMode = Literal["manual", "none"]


class OperationalResolutionRequest(BaseModel):
    """Request body for ``POST /operational-resolution/run``.

    Every field except ``template_id`` is optional. Posting a body
    with only the template id yields a baseline run that uses zero
    runtime context (useful for "is this template ready to receive
    operational input" checks before wiring extraction).
    """

    template_id: UUID
    pattern_id: UUID | None = None
    document_id: UUID | None = None
    batch_id: UUID | None = None

    # Accept either:
    #   - mapping shape  : {"invoice_number": "INV-1", "amount": "99.50"}
    #   - list shape     : [{"field_key": "invoice_number", ...}]
    # Both flow through ``ResolverInput.extracted_facts`` validators.
    extracted_facts: (
        list[ExtractedFact] | dict[str, Any] | None
    ) = None
    catalog_hints: dict[str, Any] | None = None
    document_metadata: dict[str, Any] | None = None
    runtime_options: dict[str, Any] | None = None

    # Pattern selection mode — explicit so future phases can add
    # ``"auto"`` / ``"hint_match"`` / etc. without breaking the
    # request shape. Today only ``"manual"`` (operator-supplied id)
    # and ``"none"`` (no pattern, raw context only) are honoured.
    pattern_selection_mode: PatternSelectionMode = "manual"

    # Hard-True for Phase 3A — protects against a future caller
    # accidentally flipping the surface into a side-effecting mode
    # before the resolver / export wiring exists.
    diagnostic_only: bool = True


# ---------------------------------------------------------------------------
# Review diagnostic (lightweight backend mirror of Phase 2G)
# ---------------------------------------------------------------------------


ReviewSeverity = Literal["ready", "info", "warning", "blocked"]
ReviewFixArea = Literal[
    "extracted_fact",
    "catalog_hint",
    "import_template",
    "invoice_pattern",
    "reference_data",
    "unknown",
]


class OperationalReviewDiagnostic(BaseModel):
    """One actionable item the operator should look at.

    Mirrors a SUBSET of the Phase 2G frontend ``PatternTestDiagnostic``
    so a future Phase 3D Review Queue surface can promote these into
    real records without re-classifying. Backend-side classification
    is intentionally narrow — the frontend Phase 2G helper covers the
    full vocabulary and stays the source of truth for nuanced
    presentation.
    """

    severity: ReviewSeverity
    code: str | None = None
    message: str
    recommendation: str | None = None
    column_id: str | None = None
    column_name: str | None = None
    fix_area: ReviewFixArea = "unknown"


# ---------------------------------------------------------------------------
# Operational summary
# ---------------------------------------------------------------------------


class OperationalResolutionSummary(BaseModel):
    """Coarse counts the operational UI / dashboards can read directly.

    Mirrors the resolver's per-row + per-severity counts, plus three
    operational aggregates (extracted fact / catalog hint count from
    the bridge input, missing-required-column count from the
    diagnostics) so a consumer can render "X of Y rows ready, Z
    required columns missing" without walking the resolver result.
    """

    status: str
    row_count: int = 0
    ready_rows: int = 0
    needs_review_rows: int = 0
    blocked_rows: int = 0
    conflict_rows: int = 0
    error_count: int = 0
    warning_count: int = 0
    info_count: int = 0
    extracted_fact_count: int = 0
    catalog_hint_count: int = 0
    missing_required_count: int = 0
    missing_fact_count: int = 0
    missing_catalog_hint_count: int = 0
    diagnostic_only: bool = True


# ---------------------------------------------------------------------------
# Top-level response envelope
# ---------------------------------------------------------------------------


class OperationalResolutionResult(BaseModel):
    """Top-level response from the operational resolution pipeline.

    Carries the resolver input + result side-by-side so a consumer
    can inspect both. ``diagnostic_only`` is hard-True at construction
    time so downstream surfaces can assert the contract at parse
    time before surfacing "this would actually export" copy.
    """

    diagnostic_only: bool = True
    template_id: str
    template_name: str | None = None
    pattern_id: str | None = None
    pattern_name: str | None = None
    document_id: str | None = None
    batch_id: str | None = None

    resolver_input: ResolverInput
    resolver_result: ResolverResult

    operational_summary: OperationalResolutionSummary
    # Empty list (not None) so consumers can iterate unconditionally.
    review_diagnostics: list[OperationalReviewDiagnostic] = Field(
        default_factory=list,
    )

File: frontend/src/types/operational-resolution.ts
/**
 * Frontend mirror of the Phase 3A backend operational-resolution
 * contract (``backend/app/schemas/operational_resolution.py``).
 *
 * The endpoint:
 *
 *   POST /api/v1/operational-resolution/run
 *
 * runs the same dry-run resolver Phase 2B uses, but in a
 * "production-shaped" envelope that carries the operational ids
 * (document_id / batch_id) Phase 3C+ will fill in for real
 * uploaded documents. Phase 3B uses it from the diagnostic
 * Operational Preview panel so operators can rehearse the
 * production flow with operator-supplied facts before OCR/AI
 * lands.
 *
 * Reuses ``ResolverInput`` / ``ResolverResult`` from
 * ``./import-resolver`` rather than re-declaring those shapes —
 * the operational pipeline is purely a composer over already-typed
 * contracts.
 *
 * Forward-compat:
 *   * ``status`` / ``severity`` / ``fix_area`` / ``pattern_selection_mode``
 *     are typed as the obvious literal sets but widened to plain
 *     ``string`` for forward compat — a future backend literal
 *     renders gracefully without crashing the panel.
 *   * Optional fields stay optional throughout.
 */

import type {
  ExtractedFact,
  ResolverInput,
  ResolverResult,
  ResolverStatus,
} from "./import-resolver";

// ---------------------------------------------------------------------------
// Request body
// ---------------------------------------------------------------------------

/**
 * Pattern selection strategy. Today only ``"manual"`` (operator
 * picks an id) and ``"none"`` (no pattern, raw context only) are
 * honoured — string-widened for future modes (``"auto"`` /
 * ``"hint_match"`` / etc.).
 */
export type PatternSelectionMode =
  | "manual"
  | "none"
  // Forward-compat: gracefully accept future literals.
  | (string & {});

/**
 * Request payload for ``POST /operational-resolution/run``.
 *
 * Every field except ``template_id`` is optional. Posting only the
 * template id yields a baseline run that uses zero runtime context
 * (useful for "is this template ready to receive operational input"
 * checks).
 */
export interface OperationalResolutionRequest {
  template_id: string;
  pattern_id?: string | null;
  document_id?: string | null;
  batch_id?: string | null;
  /**
   * Mapping shape (``{key: value}``) OR list shape
   * (``ExtractedFact[]``). Both flow through the resolver's
   * own validators on the backend side.
   */
  extracted_facts?: Record<string, unknown> | ExtractedFact[];
  /**
   * Catalog hints keyed by kind: ``vendor`` / ``property`` / ``gl``.
   * Values may be strings (treated as ``CatalogHint(text=...)``)
   * or full dicts (``entry_id`` / ``text`` / ``field_values``).
   */
  catalog_hints?: Record<string, unknown>;
  /**
   * Caller free-form metadata. Bridge / runner / operational
   * protected keys are filtered server-side before merge so a
   * caller can never spoof provenance.
   */
  document_metadata?: Record<string, unknown>;
  /**
   * Forwarded onto ``ResolverInput.runtime_options``. Reserved for
   * Phase 2+ scenario inputs.
   */
  runtime_options?: Record<string, unknown>;
  pattern_selection_mode?: PatternSelectionMode;
  /** Hard-True on the backend; provided here for future flexibility. */
  diagnostic_only?: boolean;
}

// ---------------------------------------------------------------------------
// Response shapes
// ---------------------------------------------------------------------------

/** Coarse aggregate counts mirroring the resolver's own summary. */
export interface OperationalResolutionSummary {
  status: ResolverStatus | string;
  row_count: number;
  ready_rows: number;
  needs_review_rows: number;
  blocked_rows: number;
  conflict_rows: number;
  error_count: number;
  warning_count: number;
  info_count: number;
  extracted_fact_count: number;
  catalog_hint_count: number;
  missing_required_count: number;
  missing_fact_count: number;
  missing_catalog_hint_count: number;
  diagnostic_only: boolean;
}

/** Per-item review surface — backend mirror of the Phase 2G frontend
 *  diagnostics, narrowed to the codes the operational pipeline
 *  classifies. */
export type OperationalReviewSeverity =
  | "ready"
  | "info"
  | "warning"
  | "blocked"
  | (string & {});

export type OperationalReviewFixArea =
  | "extracted_fact"
  | "catalog_hint"
  | "import_template"
  | "invoice_pattern"
  | "reference_data"
  | "unknown"
  | (string & {});

export interface OperationalReviewDiagnostic {
  severity: OperationalReviewSeverity;
  code?: string | null;
  message: string;
  recommendation?: string | null;
  column_id?: string | null;
  column_name?: string | null;
  fix_area?: OperationalReviewFixArea;
}

/**
 * Top-level response envelope for the operational resolution
 * endpoint. ``diagnostic_only`` is hard-True at the backend;
 * surface that in any "is this safe to run for real?" UI gate.
 */
export interface OperationalResolutionResult {
  diagnostic_only: boolean;
  template_id: string;
  template_name?: string | null;
  pattern_id?: string | null;
  pattern_name?: string | null;
  document_id?: string | null;
  batch_id?: string | null;
  resolver_input: ResolverInput;
  resolver_result: ResolverResult;
  operational_summary: OperationalResolutionSummary;
  review_diagnostics: OperationalReviewDiagnostic[];
}
